### DISCUSSION TITLE
datatable sort and filter

### DISCUSSION THREAD

**Discussion by rh-afk** [widgetti/solara#279] [Q&A]
Hi how to sort datatable and also filtering, search data same with tabulator?

thanks

**Comment by icduck** [comment]
Hi,
It's better to build the Tabulator Widgets so that you can utilize all tabulator features. I created both basic Gridjs Widget and Tabulator widget. However, after I added css to solara dataframe and twisted it a bit, I decided to use solara.dataframe in my app because its loading time is faster for a huge amount of data and no having much visual glitch like the others.

**Reply by rh-afk** [reply]
thanks @icduck but solara.dataframe don't have any tools for search, tick the box, etc as we see in tabulator. did you have any exaple for basic Girdjs widget dan tabulator widget. To be honest I am new with solara. Or there is any documentation in solara dev that I can use for doing that.
thanks

**Reply by icduck** [reply]
you can use ipyaggrid as well for your quick need. Hopefully it can help
https://github.com/widgetti/ipyaggrid

Below is an example with sort and filter enabled in config:

import ipyaggrid
import solara

def generate_column_defs(cars):
    return [{'field': key, 'filter': True} for key in cars[0].keys()] if cars else []

@solara.component
def Page():
    cars = [
        {"carName": "Chevelle", "origin": "US", "make": "Chevrolet", "price": 30415},
        {"carName": "Skylark 320", "origin": "US", "make": "Buick", "price": 21042},
        {"carName": "PL411", "origin": "Asia", "make": "Datsun", "price": 27676},
    ]

    grid_options = {
        "columnDefs": generate_column_defs(cars),
        "defaultColDef": {
            "sortable": True
        },
    }
    with solara.Column():
        ipyaggrid.Grid.element(grid_data=cars,grid_options=grid_options,
                                theme="ag-theme-balham-dark",quick_filter=True)


![sample_aggrid](https://github.com/widgetti/solara/assets/117786154/c1bfa01e-db46-4035-920c-e2f2546c7e2b)

**Reply by icduck** [reply]
also, you should be able to interact with this widget through our own custom javascript to call grid apis. Below snippet returns the selected row to one of my Solara reactive variable. Or, just simply set the export mode and observe grid_data_out

`
js_post_grid="""  
		gridOptions.api.addEventListener('selectionChanged', function () {
			// Get selected rows
			var selectedRows = gridOptions.api.getSelectedNodes();
			var data_selected = selectedRows.map(function (rowNode) {
				return rowNode.data;
			})
			// Print selected rows to the console
			  var return_data = {"selected_row":data_selected}
			  console.log(return_data)
			  view.model.set("user_params",return_data);
			  view.model.save_changes()  
		});
	"""

the similar approach can be done to update data grid through gridOptions.api.setRowData() if you don't want to use the built-in function: update_grid_data()
`

**Reply by rh-afk** [reply]
many thanks @icduck  for your advice, I will try to learn it. Did you test it can be combined with cross filtering (between dataframe, graph or map) as well.

thanks
